lib/handler.js

```diff
@@ -54,3 +77,105 @@ const notFound = (res) => {
   res.statusCode = 404
   res.end('Not Found')
 }
+
+/**
+ * @param {string} pathname
+ * @param {string} search
+ * @param {import('node:http').ServerResponse} res
+ * @param {string} querystring
+ * @param {import('pg').Pool} pgPool
+ */
+const getRetrievalSuccessRate = async (pathname, search, res, pgPool) => {
+  const filter = parseAndValidateFilter(pathname, search, res)
+  if (res.headersSent) return
+
+  setCacheControlForStatsResponse(res, filter)
+
+  // Fetch the "day" (DATE) as a string (TEXT) to prevent node-postgres for converting it into
+  // a JavaScript Date with a timezone, as that could change the date one day forward or back.
+  const { rows } = await pgPool.query(
+    'SELECT day::text, total, successful FROM retrieval_stats WHERE day >= $1 AND day <= $2',
+    [filter.from, filter.to]
+  )
+  const stats = rows.map(r => ({
+    day: r.day,
+    success_rate: r.total > 0 ? r.successful / r.total : null
+  }))
+  json(res, stats)
+}
+
+export const today = () => new Date().toISOString().split('T')[0]
+
+/**
+ * @param {string} pathname
+ * @param {string} search
+ * @param {import('node:http').ServerResponse} res
+ * @returns {{from: string | undefined; to: string | undefined}}
+ */
+export const parseAndValidateFilter = (pathname, search, res) => {
+  let { from, to } = qs.parse(search)
```

Comment:
can be refactored using the `Url` global here as well

Reply:
Good point. I see that Url provides `searchParams` with key/value entries.

```
> new URL('http://example.com/foo?key=value')
URL {
  href: 'http://example.com/foo?key=value',
  origin: 'http://example.com',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'example.com',
  hostname: 'example.com',
  port: '',
  pathname: '/foo',
  search: '?key=value',
  searchParams: URLSearchParams { 'key' => 'value' },
  hash: ''
}
```

Reply:
See bc6ec03